src/devices/bootloader/update.rs

```diff
@@ -78,17 +78,21 @@ impl<
     fn attempt_serial_update(&mut self) -> Option<Image<MCUF::Address>> {
         // Restore the update plan so the attempt is done only once.
         self.update_planner.as_mut().unwrap().write_update_plan(UpdatePlan::None);
-        duprintln!(
-            self.serial,
-            "Please send firmware image via XMODEM.",
-        );
-        let bank = self.boot_bank();
-        let blocks = self.serial.as_mut().unwrap().blocks(None);
-        if self.mcu_flash.write_from_blocks(bank.location, blocks).is_err() {
-            duprintln!(self.serial, "FATAL: Failed to flash image during serial update.",);
-            panic!();
+
+        let boot_bank = self.boot_bank();
+
+        if let Some(serial) = self.serial.as_mut() {
+            uprintln!(serial, "Please send firmware image via XMODEM.");
+
+            if self.mcu_flash.write_from_blocks(boot_bank.location, serial.blocks(None)).is_err() {
+                uprintln!(serial, "FATAL: Failed to flash image during serial update.");
+                panic!();
+            }
+        } else {
+            defmt::error!("Cannot perform serial update without serial console.");
```

Comment:
Isn't this going to "silently" fail with a release build?

Reply:
since there's no serial console, im not sure there IS a way to communicate this kind of failure to a user in release mode

Reply:
This isn't quite the right way to do things though. The function returned `Option` rather than `Error` because it had no error path, only two valid output scenarios (received a full image or didn't). Now there is an error path, so the function should be refactored to return `Error` instead. The decision of what to do with errors (print them or swallow them) should happen at layers above.